datadog/resource_datadog_logs_archive.go

```diff
@@ -0,0 +1,355 @@
+			return "gcs", buildGCSMap(*d), nil
+		case *datadogV2.LogsArchiveDestinationS3:
+			return "s3", buildS3Map(*d), nil
+		}
+	}
+	return "", emptyDestination, fmt.Errorf("Destination should be not null.")
+}
+
+func buildAzureMap(destination datadogV2.LogsArchiveDestinationAzure) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["client_id"] = destination.Integration.ClientId
+	result["tenant_id"] = destination.Integration.TenantId
+	result["container"] = destination.Container
+	result["storage_account"] = destination.StorageAccount
+	result["path"] = destination.Path
+	return result
+}
+
+func buildGCSMap(destination datadogV2.LogsArchiveDestinationGCS) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["client_email"] = destination.Integration.ClientEmail
+	result["project_id"] = destination.Integration.ProjectId
+	result["bucket"] = destination.Bucket
+	result["path"] = destination.Path
+	return result
+}
+
+func buildS3Map(destination datadogV2.LogsArchiveDestinationS3) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["account_id"] = destination.Integration.AccountId
+	result["role_name"] = destination.Integration.RoleName
+	result["bucket"] = destination.Bucket
+	result["path"] = destination.Path
+	return result
+}
+
+//Map to model
+func buildDatadogArchiveCreateReq(d *schema.ResourceData) (datadogV2.LogsArchiveCreateRequest, error) {
+	archive := datadogV2.NewLogsArchiveCreateRequest()
+	destination, err := buildCreateReqDestination(d)
+	if err != nil {
+		return *archive, err
+	}
+	attributes := datadogV2.NewLogsArchiveCreateRequestAttributes(
+		destination,
+		d.Get("name").(string),
+		d.Get("query").(string),
+	)
+	definition := datadogV2.NewLogsArchiveCreateRequestDefinitionWithDefaults()
+	definition.SetAttributes(*attributes)
+	archive.SetData(*definition)
+	return *archive, nil
+}
+
+func buildCreateReqDestination(d *schema.ResourceData) (datadogV2.LogsArchiveCreateRequestDestination, error) {
+	emptyDestination := datadogV2.LogsArchiveCreateRequestDestination{}
+	defDestinations := definedDestinations(d)
+	if len(defDestinations) != 1 {
+		return emptyDestination, fmt.Errorf("More than one type defined: %v", defDestinations)
+	}
+	archiveType := defDestinations[0]
+	if buildFunction, exists := buildCreateReqDestinationByTypeFunctions[archiveType]; exists {
+		return buildFunction(d.Get(archiveType).(map[string]interface{}))
+	}
+	return emptyDestination, fmt.Errorf("Archive type '%s' doesn't exist", archiveType)
+}
+
+func definedDestinations(d *schema.ResourceData) []string {
+	defined := []string{}
+	for _, destination := range []string{"azure", "gcs", "s3"} {
+		if _, ok := d.GetOk(destination); ok {
+			defined = append(defined, destination)
+		}
+	}
+	return defined
+}
+
+var buildCreateReqDestinationByTypeFunctions = map[string]func(map[string]interface{}) (datadogV2.LogsArchiveCreateRequestDestination, error){
```

Comment:
This variable is only used in one place. Do we need to keep this as a variable?

Reply:
I agree, we don't need it. I'll move that back to the only place where it is called.